File: android/app/src/main/java/wilton/calls/compat/GetWiltoncallConfig.java
/*
 * Copyright 2020, alex at staticlibs.net
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package wilton.calls.compat;

import com.google.gson.reflect.TypeToken;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.LinkedHashMap;

import wilton.Call;
import wilton.support.WiltonException;

import static org.apache.commons.io.FileUtils.readFileToString;
import static mobile.example.MainActivity.activity;
import static wilton.support.WiltonJson.wiltonFromJson;
import static wilton.support.WiltonJson.wiltonToJson;

public class GetWiltoncallConfig implements Call {

    private final Config config;

    public GetWiltoncallConfig() throws WiltonException {
        try {
            File filesDir = activity().getExternalFilesDir(null);
            String baseUrl = new File(filesDir, "stdlib").getAbsolutePath();
            LinkedHashMap<String, String> paths = new LinkedHashMap<>();
            File packagesFile = new File(filesDir, "stdlib/wilton-requirejs/wilton-packages.json");
            String text = readFileToString(packagesFile, "UTF-8");
            paths.put("example", new File(filesDir, "app").getAbsolutePath());
            ArrayList<Config.RequireJS.Package> packages = wiltonFromJson(text, Config.RequireJS.PACKAGES_JSON_TYPE);
            this.config = new Config(baseUrl, paths, packages);
        } catch (IOException e) {
            throw new WiltonException("Error retrieving Wilton config", e);
        }
    }

    public String call(final String data) throws Exception {
        return wiltonToJson(this.config);
    }

    private static class Config {
        private static final String MOBILE_UNSUPPORTED = "MOBILE_UNSUPPORTED";

        private String defaultScriptEngine = "rhino";
        private String wiltonExecutable = MOBILE_UNSUPPORTED;
        private String wiltonHome = MOBILE_UNSUPPORTED;
        private String wiltonVersion = MOBILE_UNSUPPORTED;
        private RequireJS requireJs;
        private LinkedHashMap<String, String> environmentVariables = new LinkedHashMap<>();
        private String compileTimeOS = "mobile";
        private int debugConnectionPort = -1;
        private boolean traceEnable = false;
        private String cryptCall = "";

        public Config() {
        }

        Config(String baseUrl, LinkedHashMap<String, String> paths, ArrayList<RequireJS.Package> packages) {
            this.requireJs = new RequireJS(baseUrl, paths, packages);
        }

        static class RequireJS {
            static final Type PACKAGES_JSON_TYPE = new TypeToken<ArrayList<Package>>() {}.getType();

            private int waitSeconds = 0;
            private boolean enforceDefine = true;
            private boolean nodeIdCompat = true;
            private String baseUrl;
            private LinkedHashMap<String, String> paths;
            private ArrayList<Package> packages;

            RequireJS() {
            }

            RequireJS(String baseUrl, LinkedHashMap<String, String> paths, ArrayList<Package> packages) {
                this.baseUrl = baseUrl;
                this.paths = paths;
                this.packages = packages;
            }

            static class Package {
                private String name;
                private String main;

                public Package() {
                }
            }
        }
    }
}

File: ios/ios/Wilton/Calls/Compat/GetWiltoncallConfig.swift
//
//  GetWiltoncallConfig.swift
//  ios
//
//  Created by Alexey Liverty on 6/1/20.
//  Copyright © 2020 alex. All rights reserved.
//

import Foundation

private let MOBILE_UNSUPPORTED: String = "MOBILE_UNSUPPORTED"

class GetWiltoncallConfig : Call {
    private let config: Config

    init() throws {
        do {
            let baseUrl = wiltonFilesDir.appendingPathComponent("stdlib", isDirectory: true)
            var paths = [String: String]()
            let packagesFile = baseUrl.appendingPathComponent("wilton-requirejs/wilton-packages.json")
            let text = try readFileToString(packagesFile)
            paths["example"] = wiltonFilesDir.appendingPathComponent("app").absoluteString
            let packages = try wiltonFromJson(text, [Config.RequireJS.Package].self)
            self.config = Config(baseUrl.absoluteString, paths, packages)
        } catch {
            throw WiltonException("GetWiltoncallConfig: Error retrieving Wilton config, message: [\(error)]");
        }
    }

    func call(_ data: String) throws -> String {
        return wiltonToJson(self.config)
    }

    private class Config : Codable {
        private let defaultScriptEngine: String = "rhino"
        private let wiltonExecutable: String = MOBILE_UNSUPPORTED
        private let wiltonHome: String = MOBILE_UNSUPPORTED
        private let wiltonVersion: String = MOBILE_UNSUPPORTED
        private let requireJs: RequireJS
        private let environmentVariables: [String: String] = [String: String]()
        private let compileTimeOS: String = "mobile";
        private let debugConnectionPort: Int = -1;
        private let traceEnable: Bool = false;
        private let cryptCall: String = "";

        fileprivate init(_ baseUrl: String, _ paths: [String: String], _ packages: [RequireJS.Package]) {
            self.requireJs = RequireJS(baseUrl, paths, packages);
        }

        fileprivate class RequireJS : Codable {
            private let waitSeconds: Int = 0
            private let enforceDefine: Bool = true
            private let nodeIdCompat: Bool = true
            private let baseUrl: String
            private let paths: [String: String]
            private let packages: [Package]

            fileprivate init(_ baseUrl: String, _ paths: [String: String], _ packages: [Package]) {
                self.baseUrl = baseUrl;
                self.paths = paths;
                self.packages = packages;
            }

            fileprivate class Package : Codable {
                private let name: String!
                private let main: String!
            }
        }
    }
}
